### ISSUE TITLE
[BUG] upgrade from 1.4.4 to 1.5.4 failed because of volumes keeping attached

### ISSUE THREAD

**Issue by tijder** [longhorn/longhorn#8315]
## Describe the bug

This the second time I tried updating Longhorn to 1.5.4 from 1.4.4. But again I can't get all the volumes to attach.
First the nodes didn't become ready. After setting the correct tag (as explained here https://github.com/longhorn/longhorn/issues/7212) the nodes became ready.
Directly some volumes became attached, without the pod was up. The pods that got to be attached I couldn't detach them.
Also some pods keeped  attaching. I didn't get them to attach. I tried restarting de nodes, starting de pod linked to the volume, disable selinux.

## To Reproduce

Upgrade to 1.5.4 from 1.4.4 with the kubectl commando:
```
kubectl apply -f https://raw.githubusercontent.com/longhorn/longhorn/v1.5.4/deploy/longhorn.yaml
```

## Expected behavior

Only the volumes attached that have k8s pods linked.

## Environment

 - Longhorn version: 1.5.4
 - Impacted volume (PV): 
   - pvc-a35a936b-30c3-4892-84d1-adb2ce4c515a
   - pvc-c488336c-29ff-4dbe-8f9c-7c16ab5b2008
   - pvc-4d501d7e-43dd-4cbc-9341-dbb0a4403c5c
   - pvc-c488336c-29ff-4dbe-8f9c-7c16ab5b2008
   - pvc-0e68938d-3762-4d12-b8f5-0de9fd19bfa1
   - pvc-637564a8-270c-4862-bf85-deeefd12a8cb
   - pvc-f3b6b2de-dcaf-41a3-b019-a058aed1b58c
 - Installation method (e.g. Rancher Catalog App/Helm/Kubectl): Kubectl
 - Kubernetes distro (e.g. RKE/K3s/EKS/OpenShift) and version: v1.28.8+k3s1
   - Number of control plane nodes in the cluster: 3
   - Number of worker nodes in the cluster: 2
 - Node config
   - OS type and version: AlmaLinux 9.3 (Shamrock Pampas Cat)
   - Kernel version: 5.14.0-362.24.2.el9_3.x86_64
   - CPU per node: 6.00
   - Memory per node: 15.34Gi
   - Disk type (e.g. SSD/NVMe/HDD): HDD
   - Network bandwidth between the nodes (Gbps): 100 Mbps
 - Underlying Infrastructure (e.g. on AWS/GCE, EKS/GKE, VMWare/KVM, Baremetal): Baremetal
 - Number of Longhorn volumes in the cluster: 41

**Comment by mantissahz** [comment]
Hi @tijder, 
Could you provide the support bundle for investigating?

Did you try to scale down and scale up the workload?
Any information from `kubectl describe pod [attaching volumes pods name]` ?

**Comment by tijder** [comment]
Hi @mantissahz, thanks for your reply. I emailed the support bundle to [longhorn-support-bundle@suse.com](mailto:longhorn-support-bundle@suse.com).

I tried scaling up and down. That didn't work. It keeped attaching the volume.

I have now reverted back to 1.4.4. When I reverted back there were still three volumes keeping attaching, but when I restart the node it was tryng to attach to it stopped attaching (I also tried that when on 1.5.4). So now everything is working fine again on 1.4.4.

I don't have time to again upgrade right know. But the command on 1.4.4 gives this info:
```
Name:             forgejo-db-598dd8867b-2xtsq
Namespace:        forgejo
Priority:         0
Service Account:  default
Node:             maloof/10.0.64.70
Start Time:       Sun, 07 Apr 2024 19:40:20 +0200
Labels:           io.kompose.service=forgejo-db
                  pod-template-hash=598dd8867b
Annotations:      <none>
Status:           Running
IP:               10.42.0.205
IPs:
  IP:           10.42.0.205
Controlled By:  ReplicaSet/forgejo-db-598dd8867b
Containers:
  forgejo-db:
    Container ID:   containerd://59845270dbb759748789a9bbace27dd8e5550ff4cff3aa4784a873e5aa593425
    Image:          postgres:15.3
    Image ID:       docker.io/library/postgres@sha256:8775adb39f0db45cf4cdb3601380312ee5e9c4f53af0f89b7dc5cd4c9a78e4e8
    Port:           5432/TCP
    Host Port:      0/TCP
    State:          Running
      Started:      Sun, 07 Apr 2024 19:40:42 +0200
    Ready:          True
    Restart Count:  0
    Environment:
      POSTGRES_DB:        gitea
      POSTGRES_USER:      gitea
      POSTGRES_PASSWORD:  gitea
      PGDATA:             /var/lib/postgresql/data/pgdata
    Mounts:
      /var/lib/postgresql/data from forgejo-claim-db (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-j9qdh (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             True 
  ContainersReady   True 
  PodScheduled      True 
Volumes:
  forgejo-claim-db:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  forgejo-claim-db
    ReadOnly:   false
  kube-api-access-j9qdh:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   BestEffort
Node-Selectors:              <none>
Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:                      <none>

```

**Comment by tijder** [comment]
Hi @mantissahz, thanks for your reply. I emailed the support bundle to [longhorn-support-bundle@suse.com](mailto:longhorn-support-bundle@suse.com).

I tried scaling up and down. That didn't work. It keeped attaching the volume.

I have now reverted back to 1.4.4. When I reverted back there were still three volumes keeping attaching, but when I restart the node it was tryng to attach to it stopped attaching (I also tried that when on 1.5.4). So now everything is working fine again on 1.4.4.

I don't have time to again upgrade right know. But the command on 1.4.4 gives this info:
```
Name:             forgejo-db-598dd8867b-2xtsq
Namespace:        forgejo
Priority:         0
Service Account:  default
Node:             maloof/10.0.64.70
Start Time:       Sun, 07 Apr 2024 19:40:20 +0200
Labels:           io.kompose.service=forgejo-db
                  pod-template-hash=598dd8867b
Annotations:      <none>
Status:           Running
IP:               10.42.0.205
IPs:
  IP:           10.42.0.205
Controlled By:  ReplicaSet/forgejo-db-598dd8867b
Containers:
  forgejo-db:
    Container ID:   containerd://59845270dbb759748789a9bbace27dd8e5550ff4cff3aa4784a873e5aa593425
    Image:          postgres:15.3
    Image ID:       docker.io/library/postgres@sha256:8775adb39f0db45cf4cdb3601380312ee5e9c4f53af0f89b7dc5cd4c9a78e4e8
    Port:           5432/TCP
    Host Port:      0/TCP
    State:          Running
      Started:      Sun, 07 Apr 2024 19:40:42 +0200
    Ready:          True
    Restart Count:  0
    Environment:
      POSTGRES_DB:        ***
      POSTGRES_USER:      ***
      POSTGRES_PASSWORD:  ***
      PGDATA:             /var/lib/postgresql/data/pgdata
    Mounts:
      /var/lib/postgresql/data from forgejo-claim-db (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-j9qdh (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             True 
  ContainersReady   True 
  PodScheduled      True 
Volumes:
  forgejo-claim-db:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  forgejo-claim-db
    ReadOnly:   false
  kube-api-access-j9qdh:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   BestEffort
Node-Selectors:              <none>
Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:                      <none>

```

**Comment by tijder** [comment]
Hi @mantissahz, did you receive the support bundle at [longhorn-support-bundle@suse.com](mailto:longhorn-support-bundle@suse.com)?

**Comment by mantissahz** [comment]
Hi @tijder, 
Yes, I am looking into the SB to find some clues.

**Comment by mantissahz** [comment]
Hi @tijder
I did not see the upgrade path from v1.4.4 to v1.5.4 in the longhorn-manager-xxxx logs.
Did you restart the longhorn-manager pods during the upgrade?
I saw some logs in the longhron-manager pod

```txt
"Failed to sync Longhorn volume" func=controller.handleReconcileErrorLogging file="utils.go:68" Volume=longhorn-system/pvc-a35a936b-30c3-4892-84d1-adb2ce4c515a controller=longhorn-volume error="failed to sync longhorn-system/pvc-a35a936b-30c3-4892-84d1-adb2ce4c515a: failed to reconcile volume state for pvc-a35a936b-30c3-4892-84d1-adb2ce4c515a: failed to find instance manager for replica pvc-a35a936b-30c3-4892-84d1-adb2ce4c515a-r-2bef9dc5: cannot find the only available instance manager for instance pvc-a35a936b-30c3-4892-84d1-adb2ce4c515a-r-2bef9dc5, node pxxxxi01, instance manager image longhornio/longhorn-instance-manager:v1.5.4, type aio" node=cancellatum
```

but I did not see the logs of pvc-a35a936b-30c3-4892-84d1-adb2ce4c515a-r-2bef9dc5 and pvc-a35a936b-30c3-4892-84d1-adb2ce4c515a-e-2bef9dc5, they seemed to be deleted.

And I can not reproduce this issue by steps:

1. Fresh install Longhorn v1.4.4.
2. Create two deployment workloads(ex: volumes v1 and v2) and wait for the workload running.
3. Scale down one of two workloads (ex volume v2 becomes detached.)
4. Upgrade Longhorn to v1.5.4 and wait for all pods running.
5. Scale up the workload (volume v2) and wait for the workload running
6. Scale down all workloads and scale up all workloads.
7. Volumes are all attached.

Could you create the support bundle when you upgrade the Longhorn to 1.5.x and find this happens again?

**Comment by tijder** [comment]
Thanks for looking in to this! I will try it again Sunday. Then send the support bundle again.
